File: backend/lib/achievements.js
const db = require('../db');
const { STREAK_MILESTONES, LEVEL_MILESTONES } = require('./leveling');

const insertIfNew = db.prepare(`
  INSERT OR IGNORE INTO achievements (user_id, code, unlocked_at) VALUES (?, ?, ?)
`);
const hasCode = db.prepare('SELECT 1 FROM achievements WHERE user_id = ? AND code = ?');

// Проверяет стрик и уровень пользователя и открывает новые бейджи.
// Возвращает список кодов, которые были открыты именно сейчас (для тостов).
function checkAndUnlock(user) {
  const unlockedNow = [];
  const now = new Date().toISOString();

  for (const days of STREAK_MILESTONES) {
    const code = `${days}d`;
    if (user.streak_current >= days && !hasCode.get(user.id, code)) {
      insertIfNew.run(user.id, code, now);
      unlockedNow.push(code);
    }
  }

  for (const lvl of LEVEL_MILESTONES) {
    const code = `lvl${lvl}`;
    if (user.level >= lvl && !hasCode.get(user.id, code)) {
      insertIfNew.run(user.id, code, now);
      unlockedNow.push(code);
    }
  }

  return unlockedNow;
}

const ACHIEVEMENT_META = {
  '7d': { icon: '🔥', name: '7 ДНЕЙ', desc: '7 дней подряд без пропусков' },
  '30d': { icon: '💪', name: '30 ДНЕЙ', desc: '30 дней подряд без пропусков' },
  '180d': { icon: '🚀', name: '180 ДНЕЙ', desc: '180 дней подряд без пропусков' },
  '365d': { icon: '👑', name: '365 ДНЕЙ', desc: 'Целый год без пропусков' },
  'lvl25': { icon: '⭐', name: 'УРОВЕНЬ 25', desc: 'Четверть пути пройдена' },
  'lvl50': { icon: '🌟', name: 'УРОВЕНЬ 50', desc: 'Половина пути пройдена' },
  'lvl75': { icon: '✨', name: 'УРОВЕНЬ 75', desc: 'Три четверти пути' },
  'lvl100': { icon: '💎', name: 'УРОВЕНЬ 100', desc: 'Version 2.0 полностью собрана' }
};

module.exports = { checkAndUnlock, ACHIEVEMENT_META };

File: internal/achievements/achievements.go
// Package achievements — разблокировка бейджей. Порт backend/lib/achievements.js.
package achievements

import (
	"strconv"
	"time"

	"version20/internal/leveling"
	"version20/internal/models"
)

type Meta struct {
	Code string
	Icon string
	Name string
	Desc string
}

var Metas = []Meta{
	{"7d", "🔥", "7 ДНЕЙ", "7 дней подряд без пропусков"},
	{"30d", "💪", "30 ДНЕЙ", "30 дней подряд без пропусков"},
	{"180d", "🚀", "180 ДНЕЙ", "180 дней подряд без пропусков"},
	{"365d", "👑", "365 ДНЕЙ", "Целый год без пропусков"},
	{"lvl25", "⭐", "УРОВЕНЬ 25", "Четверть пути пройдена"},
	{"lvl50", "🌟", "УРОВЕНЬ 50", "Половина пути пройдена"},
	{"lvl75", "✨", "УРОВЕНЬ 75", "Три четверти пути"},
	{"lvl100", "💎", "УРОВЕНЬ 100", "Version 2.0 полностью собрана"},
}

func MetaByCode(code string) Meta {
	for _, m := range Metas {
		if m.Code == code {
			return m
		}
	}
	return Meta{Code: code}
}

type checker interface {
	HasAchievement(userID int64, code string) (bool, error)
	InsertAchievement(userID int64, code, unlockedAt string) error
}

// CheckAndUnlock проверяет стрик и уровень пользователя и открывает новые
// бейджи. Возвращает список кодов, открытых именно сейчас.
func CheckAndUnlock(c checker, u *models.User) ([]string, error) {
	var unlocked []string
	now := time.Now().UTC().Format(time.RFC3339)

	type milestone struct {
		code    string
		reached bool
	}
	var all []milestone
	for _, d := range leveling.StreakMilestones {
		all = append(all, milestone{code: strconv.Itoa(d) + "d", reached: u.StreakCurrent >= d})
	}
	for _, l := range leveling.LevelMilestones {
		all = append(all, milestone{code: "lvl" + strconv.Itoa(l), reached: u.Level >= l})
	}

	for _, m := range all {
		if !m.reached {
			continue
		}
		has, err := c.HasAchievement(u.ID, m.code)
		if err != nil {
			return nil, err
		}
		if has {
			continue
		}
		if err := c.InsertAchievement(u.ID, m.code, now); err != nil {
			return nil, err
		}
		unlocked = append(unlocked, m.code)
	}
	return unlocked, nil
}
